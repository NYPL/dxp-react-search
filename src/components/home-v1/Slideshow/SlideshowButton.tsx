import React from "react";
// Components
import { Button } from "@chakra-ui/react";

interface SlideshowButtonProps {
  direction: string;
  nextSlide?: any;
  prevSlide?: any;
  visibility?: any;
}

function SlideshowButton({
  direction,
  nextSlide,
  prevSlide,
  visibility,
}: SlideshowButtonProps) {
  return (
    <Button
      tabIndex={-1}
      variant="slide-show"
      visibility={visibility}
      //@ts-ignore
      direction={direction}
      onClick={direction === "prev" ? prevSlide : nextSlide}
    >
      {direction === "prev" ? "<" : ">"}
    </Button>
  );
}
export default SlideshowButton;
