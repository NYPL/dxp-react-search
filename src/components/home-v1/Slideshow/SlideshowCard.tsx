import React from "react";
import { useStyleConfig } from "@chakra-ui/system";
// Components
import { Heading, Text, Grid, GridItem, Box } from "@chakra-ui/react";
import HomePageLink from "../HomePageLink";
import Image from "./../../shared/Image";
import { ImageType } from "../../shared/Image";

export interface SlideshowCardItem {
  id: string;
  title: string;
  author: string;
  genre: string;
  audience: string;
  image: ImageType;
  url?: string;
}

interface SlideshowCardProps {
  item: SlideshowCardItem;
}

function SlideshowCard({ item }: SlideshowCardProps) {
  // Get Card theme styles
  const styles = useStyleConfig("Card", { variant: "slide-show-card" });
  // Generate describedBy string
  const describedByIdsArray = [];
  const omitItems = ["id", "title", "image", "url", "__typename"];
  for (const propName in item) {
    if (!omitItems.includes(propName)) {
      describedByIdsArray.push(`slideshow-card-${propName}-${item.id}`);
    }
  }
  const describedByIdsString = describedByIdsArray.join(" ");

  return (
    <Grid
      id={`slideshow-card-${item.id}`}
      data-testid={`slideshow-card-${item.id}`}
      templateRows={{ base: "1fr", md: "min-content" }}
      templateColumns={{ base: "5fr 9fr", md: "1fr" }}
      sx={styles}
    >
      <GridItem className="textBox">
        <Heading
          as="h3"
          {...(describedByIdsString && {
            "aria-describedby": describedByIdsString,
          })}
        >
          <HomePageLink
            id={`slideshow-card-heading-link-${item.id}`}
            href={item.url}
          >
            {item.title}
          </HomePageLink>
        </Heading>
        <Box className="details">
          <Text id={`slideshow-card-author-${item.id}`}>{item.author}</Text>
          <Text as="span" id={`slideshow-card-audience-${item.id}`}>
            {item.audience}
          </Text>
          <Text as="span" id={`slideshow-card-genre-${item.id}`}>
            {item.genre}
          </Text>
        </Box>
      </GridItem>
      <GridItem colStart={1} rowStart={1}>
        <HomePageLink
          id={`slideshow-card-image-link-${item.id}`}
          href={item.url}
          aria-label={`${item.title}-image`}
          tabIndex={-1}
        >
          <Image
            id={`slideshow-${item.image.id}`}
            alt={item.image.alt}
            uri={item.image.uri}
            useTransformation={true}
            transformations={item.image.transformations}
            transformationLabel={"max_width_960"}
            layout="responsive"
            width={item.image.width}
            height={item.image.height}
            quality={90}
          />
        </HomePageLink>
      </GridItem>
    </Grid>
  );
}

export default SlideshowCard;
