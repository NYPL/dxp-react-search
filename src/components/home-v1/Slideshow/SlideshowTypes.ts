import { ImageType } from "../../shared/Image";

export interface SlideshowItem {
  id: string;
  title: string;
  author: string;
  genre: string;
  audience: string;
  image: ImageType;
  url: string;
}
