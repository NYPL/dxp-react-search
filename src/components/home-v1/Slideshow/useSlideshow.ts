import React from "react";

/**
 * Custom hook that controls the sliding function for the slideshow wrapper.
 * This returns functions to use for the "previous" and "next" buttons as well
 * as a CSS style object that should be use to transition between slides.
 * Inspired by: the DS carousel hook and https://codesandbox.io/s/fxjeo
 */
const useSlideshowStyles = (slidesCount = 0, slideWidth = 100) => {
  const [currentSlide, setCurrentSlide] = React.useState(0);

  const prevSlide = () => {
    setCurrentSlide((slide) => (slide === 0 ? 0 : slide - 1));
  };

  const nextSlide = () => {
    setCurrentSlide((slide) =>
      slide === slidesCount - 1 ? slidesCount - 1 : slide + 1
    );
  };
  /* Updates the left margin for the slideshow so the elements can slide inside
   the container. */
  const slideshowStyle = {
    transition: "all .5s",
    marginLeft: `-${currentSlide * slideWidth}%`,
  };
  return { currentSlide, prevSlide, nextSlide, slideshowStyle };
};

export default useSlideshowStyles;
