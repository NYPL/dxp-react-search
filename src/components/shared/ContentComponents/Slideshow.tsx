import React from "react";
import { Box } from "@nypl/design-system-react-components";
interface SlideshowProps {
  id: string;
  type: string;
  heading?: string;
  description?: string;
}

function Slideshow({ id, type, heading, description }: SlideshowProps) {
  return (
    <Box id={`${type}-${id}`} mb="xl">
      <h3>{type}</h3>
      {id}
    </Box>
  );
}

export default Slideshow;
