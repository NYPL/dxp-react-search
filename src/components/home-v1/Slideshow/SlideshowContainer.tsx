import React from "react";
// Components
import { Box, Grid, GridItem } from "@chakra-ui/react";
import Card from "../CardGrid/Card";
// Types
import { SlideshowItem } from "./SlideshowTypes";

interface SlideshowContainerProps {
  items: SlideshowItem[];
  slideshowStyle: Record<string, string>;
  currentSlide: any;
  nextSlide: () => void;
  prevSlide: () => void;
}
function SlideshowContainer({
  items,
  slideshowStyle,
  currentSlide,
  nextSlide,
  prevSlide,
}: SlideshowContainerProps) {
  const changeSlide = (i: number) => {
    if (currentSlide === 0 && i === 0) return;
    else if (currentSlide < i + 1) {
      nextSlide();
    } else {
      prevSlide();
    }
  };
  return (
    <Box w="full" overflow="hidden">
      <Grid
        as="ul"
        h={{ base: "310px", md: "300px", lg: "340px" }}
        w="full"
        listStyleType="none"
        templateColumns="repeat(10, 1fr)"
        gap={{ base: 4, md: 5, lg: "32px" }}
        ml={slideshowStyle.marginLeft}
        transition={slideshowStyle.transition}
      >
        {items &&
          items.map((item: SlideshowItem, i) => {
            return (
              <GridItem
                as="li"
                key={item.id}
                h="full"
                onFocus={() => changeSlide(i)}
              >
                <Card item={item} variant="slide-show-card" />
              </GridItem>
            );
          })}
      </Grid>
    </Box>
  );
}
export default SlideshowContainer;
