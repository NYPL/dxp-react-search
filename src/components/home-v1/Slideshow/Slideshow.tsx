import React from "react";
// Components
import { Box } from "@chakra-ui/react";
import ComponentWrapper, { SeeMore } from "../ComponentWrapper";
import SlideshowContainer from "./SlideshowContainer";
import SlideshowButton from "./SlideshowButton";
// Hooks
import useSlideshowStyles from "./useSlideshow";
// Types
import { SlideshowItem } from "./SlideshowTypes";

interface SlideshowProps {
  id?: string;
  title: string;
  link: string;
  items: SlideshowItem[];
  seeMore?: SeeMore;
}

function Slideshow({ id, title, link, items, seeMore }: SlideshowProps) {
  const { currentSlide, prevSlide, nextSlide, slideshowStyle } =
    useSlideshowStyles(items.length, 15.7);
  return (
    <ComponentWrapper
      id={id}
      title={title}
      link={link}
      textColor="brand.100"
      bg="red.100"
      borderColor="red.100"
      buttonBorder="brand.100"
      seeMore={seeMore}
    >
      <Box
        w="full"
        position="relative"
        mt={{ base: `${currentSlide === 0 ? 0 : -9}`, md: 0 }}
      >
        {currentSlide > 0 && (
          <SlideshowButton direction={"prev"} prevSlide={prevSlide} />
        )}
        <SlideshowContainer
          items={items}
          slideshowStyle={slideshowStyle}
          currentSlide={currentSlide}
          nextSlide={nextSlide}
          prevSlide={prevSlide}
          sectionTitle={title}
        />
        <SlideshowButton
          direction={"next"}
          nextSlide={nextSlide}
          visibility={currentSlide !== items.length - 1 ? "visibile" : "hidden"}
        />
      </Box>
    </ComponentWrapper>
  );
}

export default Slideshow;
