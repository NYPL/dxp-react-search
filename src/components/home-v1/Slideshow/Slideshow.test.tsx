import * as React from "react";
import { jest } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { axe, toHaveNoViolations } from "jest-axe";
import "@testing-library/jest-dom/extend-expect";
expect.extend(toHaveNoViolations);
import renderer from "react-test-renderer";
// Component
import Slideshow from "./Slideshow";
import SlideshowContainer from "./SlideshowContainer";
import SlideshowButton from "./SlideshowButton";
import useSlideshowStyles from "./useSlideshow";

// Mock useSlideshow hook
// jest.mock("./useSlideshowStyle", () => {
//   const { currentSlide, prevSlide, nextSlide, slideshowStyle } =
//     useSlideshowStyles(items.length, 11);
//   return {
//     useSlideshowStyles: () => {
//       return { currentSlide, prevSlide, nextSlide, slideshowStyle };
//     },
//   };
// });

// const mockUseSlideshowStyle = useSlideshowStyles as jest.MockedFunction<
//   typeof useSlideshowStyles
// >;

const items = [
  {
    id: "test-id-1",
    title: "Test 1",
    author: "writer 1",
    genre: "genre",
    audience: "people",
    image: "https://placeimg.com/400/200/arch",
    url: "https://www.nypl.org/",
  },
  {
    id: "test-id-2",
    title: "Test 2",
    author: "writer 2",
    genre: "genre",
    audience: "people",
    image: "https://placeimg.com/400/200/arch",
    url: "https://www.nypl.org/",
  },
  {
    id: "test-id-3",
    title: "Test 3",
    author: "writer 3",
    genre: "genre",
    audience: "people",
    image: "https://placeimg.com/400/200/arch",
    url: "https://www.nypl.org/",
  },
  {
    id: "test-id-4",
    title: "Test 4",
    author: "writer 4",
    genre: "genre",
    audience: "people",
    image: "https://placeimg.com/400/200/arch",
    url: "https://www.nypl.org/",
  },
];

xdescribe("SlideshowButton tests", () => {
  // This is not working!
  let mockPrevSlide: jest.MockedFunction<() => void>;
  let mockNextSlide: jest.MockedFunction<() => void>;
  beforeEach(() => {
    mockPrevSlide = jest.fn();
    mockNextSlide = jest.fn();
  });

  it("should pass axe accessibility test", async () => {
    const { container } = render(<SlideshowButton direction="next" />);
    expect(await axe(container)).toHaveNoViolations();
  });
  it("should render correct button according to direction prop", () => {
    render(<SlideshowButton direction="next" />);
    expect(screen.getByRole("button", { name: />/i })).toBeInTheDocument();
  });
  it("should call correct function when clicked", () => {
    render(
      <SlideshowButton
        direction="next"
        nextSlide={mockNextSlide}
        prevSlide={mockPrevSlide}
      />
    );
    userEvent.click(screen.getByRole("button", { name: />/i }));
    expect(mockNextSlide).toBeCalledTimes(1);
    expect(mockPrevSlide).not.toBeCalled();
  });
  it("should render the UI snapshot correctly", () => {
    const nextButton = renderer
      .create(<SlideshowButton direction="next" nextSlide={mockNextSlide} />)
      .toJSON();
    const prevButton = renderer
      .create(<SlideshowButton direction="prev" prevSlide={mockPrevSlide} />)
      .toJSON();
    expect(nextButton).toMatchSnapshot();
    expect(prevButton).toMatchSnapshot();
  });
});

xdescribe("SlideshowContainer tests", async () => {
  //   it("should pass axe accessibility test", async () => {
  //     const { container } = render(<SlideshowContainer items={items} />);
  // expect(await axe(container)).toHaveNoViolations();
  //   });
  it("should render list of cards ", () => {});
  it("should render the UI snapshot correctly", () => {});
});

xdescribe("Slideshow tests", () => {
  it("shold pass axe accessibility test", async () => {
    const { container } = render(
      <Slideshow title="Test" link="https://nypl.com" items={items} />
    );
    expect(await axe(container)).toHaveNoViolations();
  });
  it("should not show the prev button upon rendering", () => {
    render(<Slideshow title="Test" link="https://nypl.com" items={items} />);
    expect(
      screen.queryByRole("button", { name: /</i })
    ).not.toBeInTheDocument();
  });
  it("should show the prev button ones the next button has been clicked", () => {
    render(<Slideshow title="Test" link="https://nypl.com" items={items} />);
    userEvent.click(screen.getByRole("button", { name: />/i }));
    expect(screen.getByRole("button", { name: /</i })).toBeInTheDocument();
  });
  // @QUESTION can this be tested? does not show margin value/ can't access position
  xit("should move the SlideshowContainer to the left if the next button is clicked", () => {
    render(<Slideshow title="Test" link="https://nypl.com" items={items} />);
    const prevPosition = screen
      .getByRole("heading", { name: /Test 1/i })
      .getBoundingClientRect();
    console.log(prevPosition);
    userEvent.click(screen.getByRole("button", { name: />/i }));
    const nextPosition = screen
      .getByRole("heading", { name: /Test 1/i })
      .getBoundingClientRect();
    console.log(nextPosition);
  });
  // @QUESTION can this be tested? does not show margin value/ can't access position
  xit("should move the SlideshowContainer to the right if the prev button is clicked", () => {});
  it("should hide the next button when the end of the list is reached", () => {
    render(<Slideshow title="Test" link="https://nypl.com" items={items} />);
    userEvent.click(screen.getByRole("button", { name: />/i }));
    userEvent.click(screen.getByRole("button", { name: />/i }));
    userEvent.click(screen.getByRole("button", { name: />/i }));
    //renders prev button
    expect(screen.getByRole("button", { name: /</i })).toBeInTheDocument();
    //does not render next button
    expect(
      screen.queryByRole("button", { name: />/i })
    ).not.toBeInTheDocument();
  });
  it("should render the UI snapshot correctly", () => {
    const basicView = renderer
      .create(<Slideshow title="Test" link="https://nypl.com" items={items} />)
      .toJSON();
    expect(basicView).toMatchSnapshot();
  });
});
