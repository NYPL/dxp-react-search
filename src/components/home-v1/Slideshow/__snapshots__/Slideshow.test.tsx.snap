// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slideshow tests should render the UI snapshot correctly 1`] = `
<div
  className="css-1kbeyk8"
>
  <div
    className="css-1fqkwup"
  >
    <div
      className="css-1m8gjtz"
    >
      <div
        className="css-y9cf5o"
      >
        <h2
          className="chakra-heading css-m35uou"
        >
          <a
            className="chakra-link css-0"
            href="https://nypl.com"
          >
            Test
          </a>
        </h2>
      </div>
    </div>
    <div
      className="css-1y2jlch"
    >
      <div
        className="css-144klqr"
      >
        <div
          className="css-1ric1bm"
        >
          <ul
            className="css-1lmqp65"
          >
            <li
              className="css-1fzgmij"
              onFocus={[Function]}
            >
              <div
                className="css-pgu20y"
              >
                <div
                  className="textBox css-0"
                >
                  <h3
                    aria-describedby="test-id-1-author test-id-1-audience test-id-1-genre"
                    className="chakra-heading css-0"
                  >
                    <a
                      className="chakra-link css-0"
                      href="https://www.nypl.org/"
                    >
                      Test 1
                       
                    </a>
                  </h3>
                  <div
                    className="details css-0"
                  >
                    <p
                      className="chakra-text css-0"
                      id="test-id-1-author"
                    >
                      writer 1
                    </p>
                    <span
                      className="chakra-text css-0"
                      id="test-id-1-audience"
                    >
                      people
                    </span>
                    <span
                      className="chakra-text css-0"
                      id="test-id-1-genre"
                    >
                      genre
                    </span>
                  </div>
                </div>
                <div
                  className="css-ys3583"
                >
                  <a
                    className="chakra-link css-0"
                    href="https://www.nypl.org/"
                    tabIndex={-1}
                  >
                    <img
                      className="chakra-image css-0"
                      role="presentation"
                      src="https://placeimg.com/400/200/arch"
                    />
                  </a>
                </div>
              </div>
            </li>
            <li
              className="css-1fzgmij"
              onFocus={[Function]}
            >
              <div
                className="css-pgu20y"
              >
                <div
                  className="textBox css-0"
                >
                  <h3
                    aria-describedby="test-id-2-author test-id-2-audience test-id-2-genre"
                    className="chakra-heading css-0"
                  >
                    <a
                      className="chakra-link css-0"
                      href="https://www.nypl.org/"
                    >
                      Test 2
                       
                    </a>
                  </h3>
                  <div
                    className="details css-0"
                  >
                    <p
                      className="chakra-text css-0"
                      id="test-id-2-author"
                    >
                      writer 2
                    </p>
                    <span
                      className="chakra-text css-0"
                      id="test-id-2-audience"
                    >
                      people
                    </span>
                    <span
                      className="chakra-text css-0"
                      id="test-id-2-genre"
                    >
                      genre
                    </span>
                  </div>
                </div>
                <div
                  className="css-ys3583"
                >
                  <a
                    className="chakra-link css-0"
                    href="https://www.nypl.org/"
                    tabIndex={-1}
                  >
                    <img
                      className="chakra-image css-0"
                      role="presentation"
                      src="https://placeimg.com/400/200/arch"
                    />
                  </a>
                </div>
              </div>
            </li>
            <li
              className="css-1fzgmij"
              onFocus={[Function]}
            >
              <div
                className="css-pgu20y"
              >
                <div
                  className="textBox css-0"
                >
                  <h3
                    aria-describedby="test-id-3-author test-id-3-audience test-id-3-genre"
                    className="chakra-heading css-0"
                  >
                    <a
                      className="chakra-link css-0"
                      href="https://www.nypl.org/"
                    >
                      Test 3
                       
                    </a>
                  </h3>
                  <div
                    className="details css-0"
                  >
                    <p
                      className="chakra-text css-0"
                      id="test-id-3-author"
                    >
                      writer 3
                    </p>
                    <span
                      className="chakra-text css-0"
                      id="test-id-3-audience"
                    >
                      people
                    </span>
                    <span
                      className="chakra-text css-0"
                      id="test-id-3-genre"
                    >
                      genre
                    </span>
                  </div>
                </div>
                <div
                  className="css-ys3583"
                >
                  <a
                    className="chakra-link css-0"
                    href="https://www.nypl.org/"
                    tabIndex={-1}
                  >
                    <img
                      className="chakra-image css-0"
                      role="presentation"
                      src="https://placeimg.com/400/200/arch"
                    />
                  </a>
                </div>
              </div>
            </li>
            <li
              className="css-1fzgmij"
              onFocus={[Function]}
            >
              <div
                className="css-pgu20y"
              >
                <div
                  className="textBox css-0"
                >
                  <h3
                    aria-describedby="test-id-4-author test-id-4-audience test-id-4-genre"
                    className="chakra-heading css-0"
                  >
                    <a
                      className="chakra-link css-0"
                      href="https://www.nypl.org/"
                    >
                      Test 4
                       
                    </a>
                  </h3>
                  <div
                    className="details css-0"
                  >
                    <p
                      className="chakra-text css-0"
                      id="test-id-4-author"
                    >
                      writer 4
                    </p>
                    <span
                      className="chakra-text css-0"
                      id="test-id-4-audience"
                    >
                      people
                    </span>
                    <span
                      className="chakra-text css-0"
                      id="test-id-4-genre"
                    >
                      genre
                    </span>
                  </div>
                </div>
                <div
                  className="css-ys3583"
                >
                  <a
                    className="chakra-link css-0"
                    href="https://www.nypl.org/"
                    tabIndex={-1}
                  >
                    <img
                      className="chakra-image css-0"
                      role="presentation"
                      src="https://placeimg.com/400/200/arch"
                    />
                  </a>
                </div>
              </div>
            </li>
          </ul>
        </div>
        <button
          className="chakra-button css-r8n6x3"
          direction="next"
          onClick={[Function]}
          tabIndex={-1}
          type="button"
        >
          &gt;
        </button>
      </div>
    </div>
    <div
      className="css-1ktje31"
    >
      <div
        className="css-9cvqd6"
      >
        <a
          aria-label="see more of Test"
          className="chakra-link css-18n047l"
          href="https://nypl.com"
        >
          SEE MORE...
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`SlideshowButton tests should render the UI snapshot correctly 1`] = `
<button
  className="chakra-button css-5c2176"
  direction="next"
  onClick={[MockFunction]}
  tabIndex={-1}
  type="button"
>
  &gt;
</button>
`;

exports[`SlideshowButton tests should render the UI snapshot correctly 2`] = `
<button
  className="chakra-button css-5c2176"
  direction="prev"
  onClick={[MockFunction]}
  tabIndex={-1}
  type="button"
>
  &lt;
</button>
`;
